File: examples/iteration/kotlin/core/list.kt
package salvo.core.list

import salvo.*
import salvo.core.array.*
import salvo.core.iterator.*

fun<T> iter__2(list: List<T>): ListYield<T> {
    return ListYield(items = list, at = 0)
}

data class ListYield<T>(
    var items: List<T>,
    var at: Int,
)

fun<T> next__2(pass: ListYield<T>): Union2<T, Finished> {
    val elem = pass.items.getOrNull(pass.at)
    if (elem == null) {
        return U2_2<T, Finished>(finished())
    }
    pass.at = pass.at + 1
    return U2_1<T, Finished>(emitted(elem))
}

File: examples/iteration/rust/core/list.rs
use crate::core_array::*;
use crate::core_iterator::*;
use crate::unions::*;

pub fn iter__2<T: Clone + 'static>(list: Vec<T>) -> ListYield<T> {
    return ListYield { items: list, at: 0 };
}

#[derive(Clone, Debug)]
pub struct ListYield<T: Clone + 'static> {
    pub items: Vec<T>,
    pub at: i32,
}

pub fn next__2<T: Clone + 'static>(pass: &mut ListYield<T>) -> Union2<T, Finished> {
    let mut elem = pass.items.get((pass.at) as usize).cloned();
    if elem.is_none() {
        return Union2::<T, Finished>::U2(finished());
    }
    pass.at = pass.at + 1;
    return Union2::<T, Finished>::U1(emitted(elem.as_ref().unwrap().clone()));
}
